Documentation/Coherency.md

```diff
@@ -0,0 +1,74 @@
+
+Cons:
+- We continue to not ship bits that match official builds, except on releases.
+- We have to deal with the problem of not getting all packages for servicing releases:
+	-  Issues https://github.com/dotnet/source-build/issues/210 and https://github.com/dotnet/source-build/issues/923 are our earlier discussions.
+	-  If we turn on BuildAllConfigurations and BuildAllPackages, we will ship bits that do not exist except on RHEL machines.  If someone shares a project but not the packages directory, restoring these packages that do not officially exist will fail.
+	-  If we don't turn on BuildAllConfigurations and BuildAllPackages, we will continue to have a large number of prebuilt packages.
+
+## Build multiple versions of each repo ##
+This would mean we build CLI with its expected version of CoreFX, and core-setup with its own expected version of CoreFX.
+Pros:
+- We match the official bits very closely.
+- We don't have to deal with servicing considerations.
+
+Cons:
+- This explodes rather quickly and would have extremely long build times.
+- We have to be careful about cross-contaminating package feeds during the builds.
+
+## Store and download packages in later builds ##
+During each build, we would save source-built packages off in an Azure account or NuGet feed.  During later builds, we would restore these packages and use them in builds later in the dependency tree.
+Pros:
+- Improved build times.
+	- Fringe benefit: incremental builds would probably finally work.
+- Still lets us eliminate prebuilt packages (all packages stored are source-built).
+- We match the official bits very closely.
+
+Cons:
+- Have to be very careful to not contaminate the storage with prebuilt packages.
+- Need internet access during the build.
+- Probably does not satisfy requirements for no prebuilt binaries.
+
+# Proposed Solution #
+---------------------
+Hybrid approach: we will enable storing and retrieving packages, but allow building multiple versions of each repo completely from source.  Daily builds and dev builds will mostly use the stored packages approach while release builds will use the completely-built-from-source approach.  Doing multiple versions of repos only for release builds should cut down on the numbers of builds that are done many times since these builds should be mostly coherent except for servicing considerations.
+
+## Phase 1 ##
+After Darc merges uberclone support (https://github.com/dotnet/arcade-services/pull/291), update our version of Darc and get rid of submodules, using an uberclone instead.  This will give us many more repos than we will actually use.  For the time being, use some `PreferredRepoHashes.props` or similar to tell the build which one to use - this should change nothing in our output.  At this point we should be able to take daily build updates but will be producing a product that differs from the official release, and will have problems if we don't implement later phases before servicing.
```

Comment:
> For the time being, use some `PreferredRepoHashes.props` or similar

This implies to me that the commit hashes are hard-coded in there... is it reasonable for phase 1 to pick the highest commit in the dependency tree, the latest commit that exists in the tree, or something like that, to get automatic updates to work without more handholding?

Reply:
@mmitche was not very optimistic about a heuristic like this working but I'll put it in as an option.

Reply:
I would pick latest.  For day to day pre-servicing builds this has the best chance of working.